shap/explainers/_kernel.py

```diff
@@ -495,7 +495,13 @@ def explain(self, incoming_instance, **kwargs):
     def not_equal(i, j):
         number_types = (int, float, np.number)
         if isinstance(i, number_types) and isinstance(j, number_types):
-            return 0 if np.isclose(i, j, equal_nan=True) else 1
+            return 0 if np.allclose(i, j, equal_nan=True) else 1
+        elif hasattr(i, "dtype") and hasattr(j, "dtype"):
+            if np.issubdtype(i.dtype, np.number) and np.issubdtype(j.dtype, np.number):
+                return 0 if np.allclose(i, j, equal_nan=True) else 1
+            if np.issubdtype(i.dtype, np.bool_) and np.issubdtype(j.dtype, np.bool_):
+                return 0 if np.allclose(i, j, equal_nan=True) else 1
+            return 0 if all(i == j) else 1
```

Comment:
I tried to address https://github.com/shap/shap/pull/3983#discussion_r1947839307, but found that I actually needed a slight adjustment to the fallback for array-like inputs with `all()`. I also added custom handling for `bool_`, which seems "ok" with `np.allclose()`.